src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs

```diff
@@ -262,7 +280,8 @@ private void SetUserPreferencesMaybeAsync(LANGPREFERENCES3[] languagePreferences
         {
             if (IsForeground())
             {
-                Marshal.ThrowExceptionForHR(_textManager.SetUserPreferences4(pViewPrefs: null, pLangPrefs: languagePreferences, pColorPrefs: null));
+                var service = _serviceInitializer.GetService(CancellationToken.None);
```

Comment:
We're in a method called SetUserPreferencesMaybeAsync.  Is there a reason that method is synchronous?

Reply:
it is called from synchrouns API.

Reply:
I don't understand the name of the method then.   :)

Reply:
I see it's using `Task.Factory.StartNew`, without tracking the resulting task. Spinning off arbitrary async work without tracking it in any way is hazardous to VS shutdown. VS can crash when the CLR shuts down the app domain while threads are still running. The guidance from the Platform team is to track all your async work, such that by the time your VS Package.Dispose() method returns, there are no threads running that you're responsible for. We have a simple pattern for making this work that involves simply scheduling all async work via `AsyncPackage.JoinableTaskFactory.RunAsync`, which not only will cause your package dispose to block on completion of async work, but also avoid deadlocks if that work needs the UI thread that the package is blocking shutdown for. In addition, you can use `AsyncPackage.DisposalToken` as a cancellation token to abort the async work to avoid blocking VS shutdown for too long.